File: 3530.maximum-profit-from-valid-topological-order-in-dag.go
#
# @lc app=leetcode id=3530 lang=golang
#
# [3530] Maximum Profit from Valid Topological Order in DAG
#
# @lc code=start
func maxProfit(n int, edges [][]int, score []int) int {
    // Build adjacency list (predecessors for each node)
    predecessors := make([][]int, n)
    for _, edge := range edges {
        u, v := edge[0], edge[1]
        predecessors[v] = append(predecessors[v], u)
    }
    
    // DP with bitmask
    // dp[mask] = max profit when nodes in mask have been processed
    dp := make([]int, 1<<n)
    for i := range dp {
        dp[i] = -1
    }
    dp[0] = 0
    
    for mask := 0; mask < (1 << n); mask++ {
        if dp[mask] == -1 {
            continue
        }
        
        // Count number of nodes already processed (1-based position for next node)
        position := 0
        for i := 0; i < n; i++ {
            if mask & (1 << i) != 0 {
                position++
            }
        }
        position++ // Convert to 1-based position for next node
        
        // Try adding each unprocessed node
        for node := 0; node < n; node++ {
            if mask & (1 << node) != 0 {
                continue // already processed
            }
            
            // Check if all predecessors are processed
            canAdd := true
            for _, pred := range predecessors[node] {
                if mask & (1 << pred) == 0 {
                    canAdd = false
                    break
                }
            }
            
            if canAdd {
                newMask := mask | (1 << node)
                newProfit := dp[mask] + score[node] * position
                if dp[newMask] < newProfit {
                    dp[newMask] = newProfit
                }
            }
        }
    }
    
    return dp[(1<<n)-1]
}
# @lc code=end

File: 3530.maximum-profit-from-valid-topological-order-in-dag.java
#
# @lc app=leetcode id=3530 lang=java
#
# [3530] Maximum Profit from Valid Topological Order in DAG
#

# @lc code=start
class Solution {
    public int maxProfit(int n, int[][] edges, int[] score) {
        int N = 1 << n;
        int[] dp = new int[N];
        dp[0] = 0;
        int[] pred_mask = new int[n];
        for (int[] edge : edges) {
            pred_mask[edge[1]] |= (1 << edge[0]);
        }
        for (int mask = 0; mask < N; mask++) {
            if (mask != 0 && dp[mask] == 0) continue;
            int pos = Integer.bitCount(mask) + 1;
            for (int v = 0; v < n; v++) {
                if ((mask & (1 << v)) == 0 && (pred_mask[v] & mask) == pred_mask[v]) {
                    int next_mask = mask | (1 << v);
                    dp[next_mask] = Math.max(dp[next_mask], dp[mask] + score[v] * pos);
                }
            }
        }
        return dp[N - 1];
    }
}
# @lc code=end